### PR TITLE
Rewrite RCNNTarget to only use tensorflow.

### PR DESCRIPTION
**Pull request by IanTayler**

### CODE REVIEW

File: luminoth/models/fasterrcnn/rcnn_target.py [hunk 1]
```diff
@@ -76,97 +68,164 @@ def _build(self, proposals, gt_boxes):
...
+        def disable_some_fgs():
+            # We want to delete a randomly-selected subset of fg_inds of
+            # size `fg_inds.shape[0] - max_fg`.
+            # We shuffle along the dimension 0 and then we get the first
+            # num_fg_inds - max_fg indices and we disable them.
+            if self._debug:
+                shuffled_inds = tf.random_shuffle(fg_inds, seed=0)
+            else:
+                shuffled_inds = tf.random_shuffle(fg_inds)
+            disable_place = (tf.shape(fg_inds)[0] - max_fg)
```
Comment (vierja) [inline luminoth/models/fasterrcnn/rcnn_target.py]: This looks like it's assuming `num_fg_inds > max_fg`.

Comment (vierja) [inline luminoth/models/fasterrcnn/rcnn_target.py]: Are there any tests for this case?

File: luminoth/models/fasterrcnn/rcnn_target.py [hunk 2]
```diff
@@ -76,97 +68,161 @@ def _build(self, proposals, gt_boxes):
...
-        overlaps_best_label = overlaps.argmax(axis=1)
+        #  if max(iou) < config.background_threshold_low then we ignore.
+        #  elif max(iou) <= config.background_threshold_high then we label
+        #      background.
+        #  elif max(iou) > config.foreground_threshold then we label with
+        #      the highest IoU in overlap.
+        #
+        # max_overlaps gets, for each proposal, the index in which we can
+        # find the gt_box with which it has the highest overlap.
+        max_overlaps = tf.reduce_max(overlaps, axis=[1])
```
Comment (IanTayler) [inline luminoth/models/fasterrcnn/rcnn_target.py]: There isn't. I'll change it to `axis=1` to avoid confusion.

File: luminoth/models/fasterrcnn/rcnn_target.py (hunk 1 above)
Comment (IanTayler) [inline luminoth/models/fasterrcnn/rcnn_target.py]: It should be the case that `num_fg_inds > max_fg` because `disable_some_fgs` only runs if the condition `tf.greater(tf.shape(fg_inds)[0], max_fg)` evaluates to true.

It may warrant adding
```
with tf.control_dependencies([tf.assert_non_negative(disable_inds, message="something")]):
    disable_inds = shuffled_inds[:disable_place]
```

File: luminoth/models/fasterrcnn/rcnn_target.py [hunk 3]
```diff
@@ -76,97 +68,161 @@ def _build(self, proposals, gt_boxes):
...
 
-        # We are going to label each proposal based on the IoU with `gt_boxes`.
-        # Start by filling the labels with -1, marking them as ignored.
-        proposals_label = np.empty((proposals.shape[0], ), dtype=np.float32)
-        proposals_label.fill(-1)
-
+        # We are going to label each proposal based on the IoU with
+        # `gt_boxes`. Start by filling the labels with -1, marking them as
+        # ignored.
+        proposals_label_shape = tf.gather(tf.shape(proposals), [0])
```
Comment (IanTayler) [inline luminoth/models/fasterrcnn/rcnn_target.py]: The difference seems to be that `[tf.shape(proposals)[0]]` gives a python `list` of `Tensor`s. (While `Tensor.shape` returns a python `Object`, `tf.shape` can't, because it runs on the graph. So it returns a `Tensor`). On the other hand, `tf.gather(tf.shape(proposals), [0])` returns a single `Tensor`.

Surprisingly enough, `tf.stack([tf.shape(proposals)[0]])` doesn't work either, so there may be more to this than meets the eye.
